keps/sig-node/4939-tls-in-grpc-probe/README.md

```diff
@@ -0,0 +1,676 @@
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+Plan is to add unit tests to `pkg/probe/grpc/grpc_test.go`
+that can verify the flags are interpreted correctly
+and that a TLS (with no verify) and non-TLS configurations work as expected.
+
+- `k8s.io/kubernetes/pkg/probe/grpc`: TBD
+
+##### Integration tests
+
+N/A, only unit tests and e2e coverage.
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+Tests in test/e2e/common/node/container_probe.go:
+- should not be restarted with a GRPC liveness probe: TBD
+- should be restarted with a GRPC liveness probe: TBD
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implement the feature.
+- Add unit and e2e tests for the feature.
+
+#### Beta
+
+- Solicit feedback from the Alpha.
+- Ensure tests are stable and passing.
+
+#### GA
+- [ ] Address feedback from beta usage
+- [ ] Validate that API is appropriate for users
+- [ ] Close on any remaining open issues & bugs
+
+### Upgrade / Downgrade Strategy
+
+Upgrade: default values for new configurables should default to the current state so upgrade should not require anything.
+
+Downgrade: gRPC probes will only support TLS in a downgrade from alpha
+
+### Version Skew Strategy
+
+Ths issue here is when nodes are behind the control plane,
+the older nodes will receive the `tls` config but ignore it which would
+cause probes to fail if TLS is required.
+
+We may not be able to graduate this widely until all kubelet version
```

Comment:
Can you add more detail here? What happens when a pod is created with the new field and deployed against
1. Older k8s release without this feature.
2. New API server with this feature enabled and older kubelet without this feature. 
3. Feature gate turned on in API server but not in the kubelet.

Reply:
sure i can try to update this tonight

Reply:
> until all kubelet version skew supports the new `tls` configuration.

Need to clarify what the Kubelet does if it receives a pod that has a probe with the TLS configuration, but the feature gate is turned off. If the Kubelet ignores the field when the feature gate is disabled, then it will  "never" be within the version skew range.

Reply:
tried to reword this section a bit, let me know if that clears it up!

@tallclair @mrunalp